### ISSUE TITLE
CS8603 False positive when using yield return with netstandard2.0

### ISSUE THREAD

**Issue by sandrohanea** [dotnet/roslyn#80867]
**Version Used**: netstandard2.0

**Steps to Reproduce**:

1. Use the following repro class in a class library:
```
public static class ReproClass
{
    public static IEnumerable<string> GetSomeStrings()
    {
        for (var length = 1; length < 10; length++)
        {
            var s= length % 2 == 0 ? null : string.Join("", Enumerable.Range(0, length).Select(i => 'A' + i));
            if (!string.IsNullOrEmpty(s))
            {
                yield return s;
            }
        }
    }
}
```
2. Add TargetFrameworks as following to the class library:
 <TargetFrameworks>netstandard2.0;netstandard2.1;net9;net10.0</TargetFrameworks>
       
3. Build

**Expected Behavior**:

No warning is displayed

**Actual Behavior**:
For netstandard2.0, CS8603 is raised as warning, breaking builds when "-warnaserror" is used.

Build log:
```
Restore complete (0.2s)
    Determining projects to restore...
    All projects are up-to-date for restore.
    info NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  ClassLibraryReferenced netstandard2.1 succeeded (0.0s) → ClassLibraryReferenced/bin/Debug/netstandard2.1/ClassLibraryReferenced.dll
  ClassLibraryReferenced net9 succeeded (0.0s) → ClassLibraryReferenced/bin/Debug/net9/ClassLibraryReferenced.dll
    NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  ClassLibraryReferenced netstandard2.0 succeeded (0.1s) → ClassLibraryReferenced/bin/Debug/netstandard2.0/ClassLibraryReferenced.dll
    NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
    /Users/sandrohanea/Projects/tests/ConsoleAppFalsePositiveNull/ClassLibraryReferenced/ReproClass.cs(12,30): error CS8603: Possible null reference return.
  ClassLibraryReferenced net10.0 succeeded (0.1s) → ClassLibraryReferenced/bin/Debug/net10.0/ClassLibraryReferenced.dll
    NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy

Build failed with 1 error(s) in 0.5s
```
**Note:**

If attempting to build again, the warning disappear => it is present only at the first build attempt.
If you want to reproduce again, you'll need to clean and rebuild.

**Impact**:
Because of this warning, pipelines are failing and the workaround is needed.

**Workaround**:
Just replace the if with (add null-forgiving operator):
```
if (!string.IsNullOrEmpty(s))
{
        yield return s!;
}
```

This shouldn't be needed, as the string.IsNullOrEmpty is already ensuring that the string is not null. (and working for other target frameworks)

**Comment by RikkiGibson** [comment]
This appears to be by-design, `string.IsNullOrEmpty` is not annotated on netstandard2.0 and likely not on netstandard2.1 either, so the compiler doesn't understand that the argument is non-null when the method returns false.

I recommend disabling nullable warnings on netstandard, when multi-targeting both net and netstandard, e.g. with

```xml
<PropertyGroup Condition="$(TargetFramework) == 'netstandard2.0'"> <!-- adjust as needed -->
    <Nullable>annotations</Nullable> <!-- use nullable annotations, and include in the resulting netstandard2.0 assembly, but do not report warnings -->
</PropertyGroup>
```
